File: gunhoflash/coin-change.py
from typing import List

class Solution:
  def coinChange(self, coins: List[int], amount: int) -> int:
    NULL = -1
    fewestNumberOfCoins = [NULL] * (amount + 1)
    fewestNumberOfCoins[0] = 0

    for i in range(amount + 1):
      candidates = list(
        filter(
          lambda v: v != NULL,
          map(
            lambda v: fewestNumberOfCoins[i - v],
            filter(
              lambda v: v <= i,
              coins
            )
          )
        )
      )

      if len(candidates):
        fewestNumberOfCoins[i] = min(candidates) + 1

    return fewestNumberOfCoins[amount]

File: gunhoflash/coin-change.ts
const NULL = -1;

function coinChange(coins: number[], amount: number): number {
  const fewestNumberOfCoins = Array(amount + 1).fill(NULL);

  fewestNumberOfCoins[0] = 0;

  for (let i = 0; i <= amount; i++) {
    const candidates = coins
      .filter(v => v <= i)
      .map(v => fewestNumberOfCoins[i - v])
      .filter(v => v !== NULL);

    if (candidates.length) {
      fewestNumberOfCoins[i] = Math.min(...candidates) + 1;
    }
  }

  return fewestNumberOfCoins[amount];
}
